src/Tmds.DBus/SignalMatchRule.cs

```diff
@@ -1,27 +1,147 @@
 // See COPYING for details
 
+using System;
+using System.Collections.Generic;
 using System.Text;
+using Tmds.DBus.Protocol;
 
 namespace Tmds.DBus
 {
-    internal class SignalMatchRule
+    /// <summary>
+    /// Match rules describe the messages that should be sent to a client,
+    /// based on the contents of the message.
+    /// </summary>
+    public sealed class SignalMatchRule
     {
+        /// <summary>
+        /// Match messages sent over or to a particular interface.
+        /// </summary>
         public string Interface { get; set; }
+
+        /// <summary>
+        /// Match messages which have the give method or signal name.
+        /// </summary>
         public string Member { get; set; }
+
+        /// <summary>
+        /// Match messages which are sent from or to the given object.
+        /// </summary>
         public ObjectPath? Path { get; set; }
 
+        /// <summary>
+        /// Match messages which are sent from or to an object for which the 
+        /// object path is either the given value, or that value followed by one
+        /// or more path components. 
+        /// </summary>
+        public ObjectPath? PathNamespace { get; set; }
+
+        /// <summary>
+        /// Arg matches are special and are used for further restricting the
+        /// match based on the arguments in the body of a message.
+        /// </summary>
+        public  (int index, string arg)[] Args { get; set; }
+
+        /// <summary>
+        /// Argument path matches provide a specialized form of wildcard
+        /// matching for path-like namespaces.
+        /// </summary>
+        public (int index, string argPath)[] ArgPaths { get; set; }
+
+        /// <summary>
+        /// Match messages whose first argument is of type STRING, and is a bus
+        /// name or interface name within the specified namespace.
+        /// </summary>
+        public string Arg0Namespace { get; set; }
+
+        internal bool MatchHeader(Header header) 
+        {
+            if (header.MessageType != MessageType.Signal)
+            {
+                return false;
+            }
+
+            if (header.Interface != Interface)
+            {
+                return false;
+            }
+
+            if (header.Member != Member)
+            {
+                return false;
+            }
+
+            if (Path.HasValue && header.Path != Path)
+            {
+                return false;
+            }
+
+            if (PathNamespace.HasValue && (!header.Path.HasValue ||
+                !header.Path.Value.Value.StartsWith(PathNamespace.Value.Value, StringComparison.Ordinal)) )
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        internal bool MatchArgs(IReadOnlyList<object> args)
+        {
+            if (Args != null)
+            {
+                foreach (var (index, arg) in Args)
+                {
+                    if ((string)args[index] != arg)
```

Comment:
The `MatchRule` tracks `strings` and we check against `objects` which leads to this weird cast.

I wonder if we shouldn't limit to first argument filtering.
`string Arg0`, `ObjectPath? Arg0Path`?

We can read it from the `MessageReader` without de-serializing the whole object and we can deserialize it once instead of per signal handler.

Reply:
I think only allowing `arg0*` for now sounds like a good idea. Are you suggesting that we should then just deserialize the 0th arg in `HandleSignal()` in that case?

Reply:
Yes.